internal_task_handlers.go

```diff
@@ -724,45 +729,123 @@ func newActivityTaskHandler(activities []activity,
 }
 
 type cadenceInvoker struct {
-	identity      string
-	service       m.TChanWorkflowService
-	taskToken     []byte
-	cancelHandler func()
-	retryPolicy   backoff.RetryPolicy
+	sync.Mutex
+	identity              string
+	service               m.TChanWorkflowService
+	taskToken             []byte
+	cancelHandler         func()
+	retryPolicy           backoff.RetryPolicy
+	heartBeatTimeoutInSec int32 // The heart beat interval configured for this activity.
+	hbBatchEndTimer       *time.Timer
+	lastProgressToReport  *[]byte
+	isInBatchingWindow    bool // Whether we started a batch of operations that need to be reported in the cycle. This gets started on a user call.
+	closeCh               chan struct{}
 }
 
 func (i *cadenceInvoker) Heartbeat(details []byte) error {
+	i.Lock()
+	defer i.Unlock()
+
+	if i.isInBatchingWindow {
+		// If we have started batching window, keep track of last reported progress.
+		i.lastProgressToReport = &details
+		return nil
+	}
+
+	nonTransientError, err := i.internalHeartBeat(details)
+
+	if err == nil || nonTransientError {
+		// We have successfully sent heartbeat, start next batching window.
+		i.lastProgressToReport = nil
+		i.isInBatchingWindow = true
+
+		// reset timer to fire before the threshold to report.
+		if i.hbBatchEndTimer != nil {
+			i.hbBatchEndTimer.Stop()
+		}
+		deadlineToTrigger := i.heartBeatTimeoutInSec
+		if deadlineToTrigger <= 0 {
+			// If we don't have any heartbeat timeout configured.
+			deadlineToTrigger = defaultHeartBeatIntervalInSec
+		}
+
+		// We set a deadline at 80% of the timeout.
+		duration := time.Duration(0.8*float32(deadlineToTrigger)) * time.Second
+		i.hbBatchEndTimer = time.NewTimer(duration)
+
+		go func() {
+			select {
+			case <-i.hbBatchEndTimer.C:
+				// We are close to deadline.
+			case <-i.closeCh:
+				// We got closed.
+				return
+			}
+
+			// We close the batch and report the progress.
+			var progressToReport *[]byte
+			i.Lock()
+			i.isInBatchingWindow = false
+			progressToReport = i.lastProgressToReport
+			i.Unlock()
+
+			if progressToReport != nil {
```

Comment:
have same question as Mihnea. If user provide no details, the progressToReport will be nil, and you are skipping the heartbeat here.